cpp/include/util/EigenUtil.hpp

```diff
@@ -206,6 +210,12 @@ auto softmax(const Tensor& tensor);
 template <typename Array>
 auto sigmoid(const Array& arr);
 
+template <typename Array>
+int argmax(const Array& arr);
+
+template <typename DerivedData, concepts::Shape1DInt DerivedIndices>
+auto slice(const Eigen::ArrayBase<DerivedData>&, const Eigen::ArrayBase<DerivedIndices>&);
```

Comment:
In my opinion, it would be clearer if you remove the `Shape1DInt` concept, as it's awkward to name a concept for the nebulous "Derived" base-class. I would rather this look like,

```
template<typename Derived1, typename Derived2>
auto slice(const Eigen::ArrayBase<Derived1>& data, const Eigen::ArrayBase<Derived2>& indices);
```

And then put a clear comment above it, explaining the type-requirements and intended usage (with an example).

The type-requirements can be added with static-assert statements in the implementation body.

Concepts are a clunky c++ feature, and there is a decent amount of overhead to read them and understand usages. For a requirement that is one-off like this, I think it is better to do it with documentation and static-asserts instead. After all, functionally, that's what using a concept in this context really is for: a compact way to encode documentation with a static-assert.

Only when the same requirement presents itself in multiple contexts, or if there are technical advantages like avoiding SFINAE complexity or achieving clean function-overloading, are concepts really warranted, I would say.